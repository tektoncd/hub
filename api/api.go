package apiserver

import (
	api "api/gen/api"
	"context"
	"log"
)

// api service example implementation.
// The example methods log the requests and return zero values.
type apisrvc struct {
	logger *log.Logger
}

// NewAPI returns the api service implementation.
func NewAPI(logger *log.Logger) api.Service {
	return &apisrvc{logger}
}

// Get all tasks and pipelines.
func (s *apisrvc) List(ctx context.Context) (err error) {
	s.logger.Print("api.list")
	return
}
