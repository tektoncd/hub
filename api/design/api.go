// Copyright Â© 2020 The Tekton Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package design

import (
	. "goa.design/goa/v3/dsl"
	cors "goa.design/plugins/v3/cors/dsl"

	// Enables the zaplogger plugin
	_ "goa.design/plugins/v3/zaplogger"
)

var _ = API("hub", func() {
	Title("Tekton Hub")
	Description("HTTP services for managing Tekton Hub")
	Version("0.1")
	Meta("swagger:example", "false")
	Server("hub", func() {
		Host("production", func() {
			URI("http://api.hub.tekton.dev")
		})

		Services(
			"admin",
			"auth",
			"category",
			"rating",
			"resource",
			"status",
			"swagger",
		)
	})

	// TODO: restrict CORS origin | https://github.com/tektoncd/hub/issues/26
	cors.Origin("*", func() {
		cors.Headers("Content-Type", "Authorization")
		cors.Methods("GET", "POST", "PUT")
	})
})
