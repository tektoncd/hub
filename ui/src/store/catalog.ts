import { Instance, types } from 'mobx-state-tree';
import { Icons } from '../common/icons';

const icons: { [catalog: string]: Icons } = {
  official: Icons.Cat,
  verified: Icons.Certificate,
  community: Icons.User
};

export const Catalog = types
  .model({
    id: types.identifierNumber,
    name: types.optional(types.string, ''),
    type: types.optional(types.string, ''),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }))
  .views((self) => ({
    get icon(): Icons {
      return icons[self.type] || Icons.Unknown;
    }
  }));

export type ICatalog = Instance<typeof Catalog>;
export type ICatalogStore = Instance<typeof CatalogStore>;

export const CatalogStore = types
  .model({
    items: types.map(Catalog)
  })

  .actions((self) => ({
    add(item: ICatalog) {
      self.items.put({ id: item.id, name: item.name, type: item.type });
    },
    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    }
  }))

  .views((self) => ({
    get values() {
      return Array.from(self.items.values());
    },

    get selected() {
      const list = new Set();
      self.items.forEach((c: ICatalog) => {
        if (c.selected) {
          list.add(c.id);
        }
      });

      return list;
    }
  }));
