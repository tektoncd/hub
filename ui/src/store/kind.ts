import { types, Instance } from 'mobx-state-tree';
import { Icons } from '../common/icons';

const icons: { [kind: string]: Icons } = {
  Task: Icons.Build,
  Pipeline: Icons.Domain
};

export const Kind = types
  .model({
    name: types.identifier,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }))
  .views((self) => ({
    get icon(): Icons {
      return icons[self.name] || Icons.Unknown;
    }
  }));

export type IKind = Instance<typeof Kind>;
export type IKindStore = Instance<typeof KindStore>;

export const KindStore = types
  .model({
    items: types.optional(types.map(Kind), {})
  })

  .actions((self) => ({
    add(item: string) {
      self.items.put({ name: item, selected: false });
    },

    clearSelected() {
      self.items.forEach((k) => {
        k.selected = false;
      });
    }
  }))

  .views((self) => ({
    get values() {
      return Array.from(self.items.values());
    },

    get selected() {
      const list = new Set();
      self.items.forEach((c: IKind) => {
        if (c.selected) {
          list.add(c.name);
        }
      });

      return list;
    }
  }));
