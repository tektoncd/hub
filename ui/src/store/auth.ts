import { flow, getEnv, Instance, types } from 'mobx-state-tree';
import { Api } from '../api';
import { titleCase } from '../common/titlecase';

export const TokenInfo = types.model('TokenInfo', {
  token: types.optional(types.string, ''),
  expiresAt: types.optional(types.number, 0),
  refreshInterval: types.optional(types.string, '')
});

export const Error = types.model('Error', {
  status: types.optional(types.number, 0),
  customMessage: types.optional(types.string, ''),
  serverMessage: types.optional(types.string, '')
});

export type ITokenInfo = Instance<typeof TokenInfo>;
export type IError = Instance<typeof Error>;

export interface AuthCodeProps {
  code: string;
}

interface Refresh {
  (): void;
}

export const AuthStore = types
  .model('AuthStore', {
    accessTokenInfo: types.optional(TokenInfo, {}),
    refreshTokenInfo: types.optional(TokenInfo, {}),
    isLoading: true,
    isAuthenticated: false,
    userRating: types.optional(types.number, 0),
    authErr: types.optional(Error, {}),
    ratingErr: types.optional(Error, {})
  })
  .actions((self) => ({
    addAccessTokenInfo(item: ITokenInfo) {
      self.accessTokenInfo.token = item.token;
      self.accessTokenInfo.expiresAt = item.expiresAt;
      self.accessTokenInfo.refreshInterval = item.refreshInterval;
    },
    addRefreshTokenInfo(item: ITokenInfo) {
      self.refreshTokenInfo.token = item.token;
      self.refreshTokenInfo.expiresAt = item.expiresAt;
      self.refreshTokenInfo.refreshInterval = item.refreshInterval;
    },
    setIsAuthenticated(l: boolean) {
      self.isAuthenticated = l;
    },
    setUserRating(rating: number) {
      self.userRating = rating;
    },
    setLoading(l: boolean) {
      self.isLoading = l;
    },
    setErrorMessage(error: IError) {
      switch (error.status) {
        case 400:
          error.customMessage = 'Bad Request';
          break;
        case 401:
          error.customMessage = 'Unauthorized. Please login again!';
          break;
        case 404:
          error.customMessage = 'Requested resource not found';
          break;
        case 500:
          error.customMessage = 'Internal server error. Please try again after some time!';
          break;
        default:
          error.customMessage = '';
      }
    },
    onFailure(err: Error) {
      self.authErr.serverMessage = err.toString();
      self.authErr.status = 400;
      self.authErr.customMessage = 'GitHub login failed';
    },
    logout() {
      localStorage.clear();
      self.isAuthenticated = false;
      self.isLoading = false;
      self.userRating = 0;
    }
  }))
  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    }
  }))
  .actions((self) => ({
    authenticate: flow(function* (authCode: AuthCodeProps, refresh?: Refresh) {
      try {
        self.setLoading(true);

        const { api } = self;
        const json = yield api.authentication(authCode.code);

        const userDetails = json.data;
        self.addAccessTokenInfo(userDetails.access);
        self.addRefreshTokenInfo(userDetails.refresh);

        self.setIsAuthenticated(true);

        if (refresh) {
          refresh();
        }
      } catch (err) {
        const error: IError = {
          status: err.status,
          serverMessage: titleCase(err.data.message),
          customMessage: ''
        };
        self.setErrorMessage(error);
        self.authErr = error;
      }
      self.setLoading(false);
    }),

    getRating: flow(function* (resourceId: number) {
      try {
        self.setLoading(true);
        const { api } = self;

        if (self.isAuthenticated) {
          const json = yield api.getRating(resourceId, self.accessTokenInfo.token);
          self.setUserRating(json.rating);
        }
      } catch (err) {
        const error: IError = {
          status: err.status,
          serverMessage: titleCase(err.data.message),
          customMessage: ''
        };
        self.setErrorMessage(error);
        self.ratingErr = error;
      }
      self.setLoading(false);
    }),

    setRating: flow(function* (resourceId: number, rating: number) {
      try {
        self.setLoading(true);

        const { api } = self;

        if (self.isAuthenticated) {
          yield api.setRating(resourceId, self.accessTokenInfo.token, rating);
          self.setUserRating(rating);
        }
      } catch (err) {
        const error: IError = {
          status: err.status,
          serverMessage: titleCase(err.data.message),
          customMessage: ''
        };
        self.setErrorMessage(error);
        self.ratingErr = error;
      }
      self.setLoading(false);
    })
  }));
export type IAuthStore = Instance<typeof AuthStore>;
