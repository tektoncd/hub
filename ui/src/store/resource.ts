import { types, Instance } from 'mobx-state-tree';
import fuzzysort from 'fuzzysort';
import moment, { Moment } from 'moment';
import { flow, getEnv } from 'mobx-state-tree';
import { Tag, ICategoryStore, ITag } from './category';
import { Api } from '../api';
import { Catalog, CatalogStore } from './catalog';
import { Kind, KindStore } from './kind';
import { assert } from './utils';

export const updatedAt = types.custom<string, Moment>({
  name: 'momentDate',
  fromSnapshot(value: string): Moment {
    return moment(new Date(value));
  },
  toSnapshot(value: Moment): string {
    return value.fromNow();
  },
  isTargetType(v: string) {
    return moment.isMoment(v);
  },
  getValidationMessage(v: string) {
    if (moment.isMoment(v)) {
      return 'Invalid moment object';
    }
    return '';
  }
});

const Version = types.model('Version', {
  id: types.identifierNumber,
  version: types.string,
  displayName: types.optional(types.string, ''),
  description: types.optional(types.string, ''),
  minPipelinesVersion: types.optional(types.string, ''),
  rawURL: types.string,
  webURL: types.string,
  updatedAt: types.optional(updatedAt, '')
});

export const Resource = types
  .model('Resource', {
    id: types.number,
    name: types.identifier,
    catalog: types.reference(Catalog),
    kind: types.reference(Kind),
    latestVersion: types.reference(Version),
    displayVersion: types.reference(Version),
    tags: types.array(types.reference(Tag)), // ["1", "2"]
    rating: types.number,
    versions: types.array(types.reference(Version)),
    displayName: ''
  })
  .views((self) => ({
    get resourceName() {
      if (self.displayName !== '') {
        return self.displayName;
      }
      return self.name;
    },
    get webURL() {
      const index = self.displayVersion.webURL.lastIndexOf('/');
      return self.displayVersion.webURL.substring(0, index + 1);
    },
    get summary() {
      const description = self.displayVersion.description;
      const index = description.indexOf('\n');
      return description.substring(0, index) || description;
    },
    get detailDescription() {
      const description = self.displayVersion.description;
      const index = description.indexOf('\n');
      return index !== -1 ? description.substring(index + 1).trim() : '';
    },
    get installCommand() {
      return `kubectl apply -f ${self.displayVersion.rawURL}`;
    }
  }));

export type IResource = Instance<typeof Resource>;
export type IVersion = Instance<typeof Version>;

export enum SortByFields {
  Unknown = '',
  Name = 'Name',
  Rating = 'Rating'
}

export const ResourceStore = types
  .model('ResourceStore', {
    resources: types.map(Resource),
    versions: types.map(Version),
    catalogs: types.optional(CatalogStore, {}),
    kinds: types.optional(KindStore, {}),
    sortBy: types.optional(types.enumeration(Object.values(SortByFields)), SortByFields.Unknown),
    tags: types.optional(types.map(Tag), {}),
    search: '',
    err: '',
    isLoading: true
  })
  .views((self) => ({
    get items() {
      return Array.from(self.resources.values());
    },
    get api(): Api {
      return getEnv(self).api;
    },
    get categories(): ICategoryStore {
      return getEnv(self).categories;
    }
  }))

  .actions((self) => ({
    setLoading(l: boolean) {
      self.isLoading = l;
    },
    setSearch(text: string) {
      self.search = text;
    },
    add(item: IResource) {
      self.resources.put(item);
    },
    setSortBy(field: string) {
      const key: SortByFields = SortByFields[field as keyof typeof SortByFields];
      self.sortBy = key;
    }
  }))

  .actions((self) => ({
    versionInfo: flow(function* (resourceName: string) {
      try {
        self.setLoading(true);

        const { api } = self;
        const resource = self.resources.get(resourceName);
        assert(resource);

        const json = yield api.resourceVersion(resource.id);

        const versions: IVersion[] = json.data.versions.map((v: IVersion) => ({
          id: v.id,
          version: v.version,
          webURL: v.webURL,
          rawURL: v.rawURL
        }));

        versions.forEach((v: IVersion) => {
          if (!self.versions.has(String(v.id))) {
            self.versions.put(v);
            if (self.resources.has(resourceName)) {
              const resource = self.resources.get(resourceName);
              assert(resource);
              resource.versions.push(v.id);
            }
          }
        });
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    }),

    versionUpdate: flow(function* (versionId: number) {
      try {
        self.setLoading(true);

        const { api } = self;
        const json = yield api.versionUpdate(versionId);

        const versionData = json.data;

        const version: IVersion = {
          id: versionData.id,
          version: versionData.version,
          displayName: versionData.displayName,
          description: versionData.description,
          minPipelinesVersion: versionData.minPipelinesVersion,
          webURL: versionData.webURL,
          rawURL: versionData.rawURL,
          updatedAt: versionData.updatedAt
        };

        self.versions.put(version);
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    }),

    load: flow(function* () {
      try {
        self.setLoading(true);

        const { api } = self;
        const json = yield api.resources();

        const kinds: string[] = json.data.map((r: IResource) => r.kind);
        kinds.forEach((k) => self.kinds.add(k));

        json.data.forEach((r: IResource) => {
          self.catalogs.add(r.catalog);
          self.versions.put(r.latestVersion);
        });

        // adding the tags to the store - normalized
        const tags: ITag[] = json.data.flatMap((item: IResource) => item.tags);

        tags.forEach((t) => (t != null ? self.tags.put(t) : null));

        const resources: IResource[] = json.data.map((r: IResource) => ({
          id: r.id,
          name: r.name,
          catalog: r.catalog.id,
          kind: r.kind,
          latestVersion: r.latestVersion.id,
          displayVersion: r.latestVersion.id,
          tags: r.tags != null ? r.tags.map((tag: ITag) => tag.id) : [],
          rating: r.rating,
          versions: [],
          displayName: r.latestVersion.displayName
        }));

        resources.forEach((r: IResource) => {
          r.versions.push(r.latestVersion);
          self.add(r);
        });
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    },
    setDisplayVersion(resourceName: string, versionId: string) {
      const resource = self.resources.get(resourceName);
      assert(resource);
      const version = self.versions.get(versionId);
      assert(version);
      if (version.id !== resource.displayVersion.id) {
        resource.displayVersion = version;
        self.versionUpdate(version.id);
      }
    }
  }))

  .views((self) => ({
    get filteredResources() {
      const { resources, kinds, catalogs, categories, search, sortBy } = self;
      const { selectedTags } = categories;

      let filteredItems: IResource[] = [];
      resources.forEach((r: IResource) => {
        const matchesKind = kinds.selected.size === 0 || kinds.selected.has(r.kind.name);
        const matchesCatalogs = catalogs.selected.size === 0 || catalogs.selected.has(r.catalog.id);
        const matchesTags = selectedTags.size === 0 || r.tags.some((t) => selectedTags.has(t.id));

        if (matchesKind && matchesCatalogs && matchesTags) {
          filteredItems.push(r);
        }
      });

      if (search.trim() !== '') {
        filteredItems = fuzzysort
          .go(search, filteredItems, { keys: ['name', 'displayName'] })
          .map((resource: Fuzzysort.KeysResult<IResource>) => resource.obj);
      }

      switch (sortBy) {
        case SortByFields.Rating:
          return filteredItems.sort((first: IResource, second: IResource) =>
            first.rating < second.rating ? 1 : first.rating > second.rating ? -1 : 0
          );

        case SortByFields.Name:
          return filteredItems.sort((first: IResource, second: IResource) =>
            first.name > second.name ? 1 : first.name < second.name ? -1 : 0
          );

        default:
          return filteredItems;
      }
    }
  }));

export type IResourceStore = Instance<typeof ResourceStore>;
