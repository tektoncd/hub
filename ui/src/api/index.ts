import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
  resourceVersion(resourceId: number): Promise<IVersion>;
  versionUpdate(versionId: number): Promise<IVersion>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resourceVersion(resourceId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async versionUpdate(versionId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/version/${versionId}`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }
}
