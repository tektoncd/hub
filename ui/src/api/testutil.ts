import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;
    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async resources() {
    const data = `${this.dataDir}/resources.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IResource>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async resourceVersion(resourceId: number) {
    const data = `${this.dataDir}/resource_versions_${resourceId}.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IVersion>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async versionUpdate(versionId: number) {
    const data = `${this.dataDir}/version_details_${versionId}.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IVersion>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
