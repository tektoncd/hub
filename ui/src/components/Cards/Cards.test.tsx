import React from 'react';
import { when } from 'mobx';
import { shallow } from 'enzyme';
import { FakeHub } from '../../api/testutil';
import { ResourceStore } from '../../store/resource';
import { CategoryStore } from '../../store/category';
import Cards from '.';

const TESTDATA_DIR = `src/store/testdata`;
const api = new FakeHub(TESTDATA_DIR);

beforeEach(() => {
  global.Date.now = jest.fn(() => new Date('2020-12-22T10:20:30Z').getTime());
});

afterEach(() => {
  global.Date = Date;
});

describe('Cards', () => {
  it('should render the resources on cards', (done) => {
    const store = ResourceStore.create({}, { api, categories: CategoryStore.create({}, { api }) });

    when(
      () => !store.isLoading,
      () => {
        const component = shallow(<Cards items={store.filteredResources} />);
        expect(component.debug()).toMatchSnapshot();

        done();
      }
    );
  });
});
