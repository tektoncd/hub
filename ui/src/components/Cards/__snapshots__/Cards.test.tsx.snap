// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards should render the resources on cards 1`] = `
"<Fragment>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"build\\" size=\\"sm\\" label=\\"Task\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"user\\" size=\\"sm\\" label=\\"tekton\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"4\\" />
            <TextContent className=\\"hub-rating\\">
              4
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            buildah
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.2
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          Buildah task builds source into a container image and then pushes it to a container registry.
          Buildah Task builds source into a container image using Project Atomic&#39;s Buildah build tool.It uses Buildah&#39;s support for building from Dockerfiles, using its buildah bud command.This command executes the directives in the Dockerfile to assemble a container image, then pushes that image to a container registry.
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              image-build
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"build\\" size=\\"sm\\" label=\\"Task\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"user\\" size=\\"sm\\" label=\\"tekton\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"5\\" />
            <TextContent className=\\"hub-rating\\">
              5
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            aws cli
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.1
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          This task performs operations on Amazon Web Services resources using aws.
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              cli
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"build\\" size=\\"sm\\" label=\\"Task\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"user\\" size=\\"sm\\" label=\\"tekton\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"5\\" />
            <TextContent className=\\"hub-rating\\">
              5
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            golang build
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.1
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          This Task is Golang task to build Go projects.
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              build-tool
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"build\\" size=\\"sm\\" label=\\"Task\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"user\\" size=\\"sm\\" label=\\"tekton\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"5\\" />
            <TextContent className=\\"hub-rating\\">
              5
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            jenkins operation
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.1
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          The following task can be used to interact with the the Jenkins REST API.
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              build
            </Badge>
            <Badge className=\\"hub-tags\\">
              jenkins
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"build\\" size=\\"sm\\" label=\\"Task\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"user\\" size=\\"sm\\" label=\\"tekton\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"5\\" />
            <TextContent className=\\"hub-rating\\">
              5
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            jib maven
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.1
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          This Task builds Java/Kotlin/Groovy/Scala source into a container image using Googleâ€™s Jib tool.
          Jib works with Maven and Gradle projects, and this template is for Maven projects.
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              image-build
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"build\\" size=\\"sm\\" label=\\"Task\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"user\\" size=\\"sm\\" label=\\"tekton\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"4.5\\" />
            <TextContent className=\\"hub-rating\\">
              4.5
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            Ansible Runner
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.1
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          Hub is a platform which displays all resources
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              cli
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
  <GalleryItem>
    <Link to={{...}} className=\\"hub-card-link\\">
      <Card className=\\"hub-resource-card\\">
        <CardHeader>
          <span className=\\"hub-kind-icon\\">
            <Icon id=\\"domain\\" size=\\"sm\\" label=\\"Pipeline\\" />
          </span>
          <span className=\\"hub-catalog-icon\\">
            <Icon id=\\"cat\\" size=\\"sm\\" label=\\"tekton-hub\\" />
          </span>
          <CardActions>
            <Icon id=\\"star\\" size=\\"sm\\" label=\\"4.5\\" />
            <TextContent className=\\"hub-rating\\">
              4.5
            </TextContent>
          </CardActions>
        </CardHeader>
        <CardTitle>
          <span className=\\"hub-resource-name\\">
            Hub Pipeline
          </span>
          <span className=\\"hub-resource-version\\">
            v
            0.1
          </span>
        </CardTitle>
        <CardBody className=\\"hub-resource-body fade\\">
          Hub is a platform which displays all resources
        </CardBody>
        <CardFooter>
          <TextContent className=\\"hub-resource-updatedAt\\">
            Updated 
            2 months ago
          </TextContent>
          <div className=\\"hub-tags-container\\">
            <Badge className=\\"hub-tags\\">
              resource
            </Badge>
          </div>
        </CardFooter>
      </Card>
    </Link>
  </GalleryItem>
</Fragment>"
`;
